pkg/controller/addon/klusterlet_addon_controller.go

```diff
@@ -0,0 +1,543 @@
+		return reconcile.Result{}, nil
+	}
+
+	nodeSelector, err := getNodeSelector(managedCluster)
+	if err != nil {
+		return reconcile.Result{}, nil
+	}
+
+	var aggregatedErrs []error
+	for addonName, enable := range agentv1.KlusterletAddons {
+		if !enable {
+			continue
+		}
+		if !addonIsEnabled(addonName, klusterletAddonConfig) {
+			if err := r.deleteManagedClusterAddon(addonName, managedCluster.GetName()); err != nil {
+				aggregatedErrs = append(aggregatedErrs, err)
+			}
+			continue
+		}
+
+		imageOverrides, err := getImageOverrides(r.client, managedCluster, addonName)
+		if err != nil {
+			return reconcile.Result{}, nil
+		}
+		gv := getGlobalValues(nodeSelector, imageOverrides, addonName, klusterletAddonConfig)
+
+		if err := r.updateManagedClusterAddon(gv, addonName, managedCluster.GetName()); err != nil {
+			aggregatedErrs = append(aggregatedErrs, err)
+		}
+	}
+	if len(aggregatedErrs) != 0 {
+		return reconcile.Result{}, fmt.Errorf("failed create/update addon %v", aggregatedErrs)
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileKlusterletAddOn) deleteAllManagedClusterAddon(clusterName string) error {
+	var aggregatedErrs []error
+	for addonName, enable := range agentv1.KlusterletAddons {
+		if !enable {
+			continue
+		}
+		if addonName == agentv1.WorkManagerAddonName {
+			continue
+		}
+		err := r.deleteManagedClusterAddon(addonName, clusterName)
+		if err != nil {
+			aggregatedErrs = append(aggregatedErrs, err)
+		}
+	}
+	if len(aggregatedErrs) != 0 {
+		return fmt.Errorf("failed to delelte all addons %v", aggregatedErrs)
+	}
+	return nil
+}
+
+func (r *ReconcileKlusterletAddOn) deleteManagedClusterAddon(addonName, clusterName string) error {
+	if !agentv1.KlusterletAddons[addonName] {
+		return nil
+	}
+	if addonName == agentv1.WorkManagerAddonName {
+		return nil
+	}
+	addon := &addonv1alpha1.ManagedClusterAddOn{}
+	err := r.client.Get(context.TODO(), types.NamespacedName{Name: addonName, Namespace: clusterName}, addon)
+	if errors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	err = r.client.Delete(context.TODO(), addon)
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+
+	return nil
+}
+
+func (r *ReconcileKlusterletAddOn) updateManagedClusterAddon(gv globalValues, addonName, clusterName string) error {
+	addon := &addonv1alpha1.ManagedClusterAddOn{}
+	err := r.client.Get(context.TODO(), types.NamespacedName{Name: addonName, Namespace: clusterName}, addon)
+	if errors.IsNotFound(err) {
+		// work manager addon cr is created by itself
+		if addonName == agentv1.WorkManagerAddonName {
+			return nil
+		}
+		newAddon := newManagedClusterAddon(addonName, clusterName)
+
+		valuesString, err := marshalGlobalValues(gv)
+		if err != nil {
+			return err
+		}
+		if len(valuesString) != 0 {
+			newAddon.SetAnnotations(map[string]string{annotationValues: valuesString})
+		}
+
+		if err := r.client.Create(context.TODO(), newAddon); err != nil {
+			return err
+		}
```

Comment:
Set `addon = newAddon` once addon is created?

Reply:
changed to` return r.client.Create(context.TODO(), newAddon). `